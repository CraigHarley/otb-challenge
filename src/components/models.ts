import {IHotel} from "../types";

export const hotels: IHotel[] = [
    {
        adults: 2,
        children: 0,
        infants: 1,
        info: "blah blah blah blah blah blah blah blah blah blah blah blah blah blah blah blah blah blah blah",
        price: "£3,136.50",
        region: "Costa Adeje, Tenerife",
        name: "Iberostar Grand Salome",
        image: "hotel-image-1.png",
        stars: 4,
        date: new Date('2019-07-03')
    },
    {
        adults: 2,
        children: 1,
        infants: 0,
        info: "blah blah blah blah blah blah blah blah blah blah blah blah blah blah blah blah blah blah blah",
        price: "£1,136.50",
        region: "Costa Adeje, Tenerife",
        name: "Wow - so fancy",
        image: "hotel-image-2.png",
        stars: 2,
        date: new Date('2019-07-03')
    },
    {
        adults: 1,
        children: 0,
        infants: 0,
        info: "blah blah blah blah blah blah blah blah blah blah blah blah blah blah blah blah blah blah blah",
        price: "£5,136.50",
        region: "Costa Adeje, Tenerife",
        name: "Swank-ville",
        image: "hotel-image-3.png",
        stars: 4,
        date: new Date('2019-07-03')
    },
    {
        adults: 2,
        children: 0,
        infants: 0,
        info: "blah blah blah blah blah blah blah blah blah blah blah blah blah blah blah blah blah blah blah",
        price: "£2,136.50",
        region: "Costa Adeje, Tenerife",
        name: "Really Fancy Hotel",
        image: "hotel-image-1.png",
        stars: 5,
        date: new Date('2019-07-03')
    }
]
