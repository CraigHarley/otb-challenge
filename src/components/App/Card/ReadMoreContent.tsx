import React, {ReactElement} from 'react';
import {css, StyleSheet} from 'aphrodite';

import config from '../../../config';

const styles = StyleSheet.create({
    readMoreContent: {
        backgroundColor: 'whitesmoke',
        width: '780px',
        margin: '0 2rem 2rem 2rem',
        padding: '1rem',
    },
    h3: {
        fontSize: 'small',
        color: config.styles.colors.primary
    },
    p: {
        fontSize: 'small',
        fontWeight: 'lighter',
        lineHeight: '1.2rem'
    },
    isHidden: {
        display: 'none'
    }
});

export interface IReadMoreContentProps {
    isShowingMoreInfo: boolean;
}

const ReadMoreContent = ({isShowingMoreInfo}: IReadMoreContentProps): ReactElement => (
    <div className={css(styles.readMoreContent, !isShowingMoreInfo && styles.isHidden)}>
        <h3 className={css(styles.h3)}>Overview</h3>
        <p className={css(styles.p)}>
            Lorem Ipsum is simply dummy text of the printing and typesetting industry. Lorem Ipsum has been the
            industry's standard dummy text ever since the 1500s, when an unknown printer took a galley of type
            and scrambled it to make a type specimen book. It has survived not only five centuries, but also the
            leap into electronic typesetting, remaining essentially unchanged. It was popularised in the 1960s with
            the release of Letraset sheets containing Lorem Ipsum passages, and more recently with desktop publishing
            software like Aldus PageMaker including versions of Lorem Ipsum.
        </p>
    </div>
)

export default ReadMoreContent;
