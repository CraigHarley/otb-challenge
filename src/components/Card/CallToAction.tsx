import React, {ReactElement} from 'react';
import {css, StyleSheet} from "aphrodite";
import config from "../../config";

const styles = StyleSheet.create({
    callToAction: {
        padding: 0,
        color: config.styles.colors.primary,
        backgroundColor: config.styles.colors.secondary,
        width: '100%',
        height: '4rem',
        border: 'none',
        borderRadius: '0.2rem',
        cursor: 'pointer',
        position: 'absolute',
        bottom: 0
    },
    actionText: {
        fontSize: 'small',
        display: 'block'
    },
    actionAmount: {
        fontWeight: 'bold',
        fontSize: 'medium'
    }
});

export interface ICallToActionProps {
    price: string
}

export default ({price}: ICallToActionProps): ReactElement =>
    <button className={css(styles.callToAction)}>
        <span className={css(styles.actionText)}>
            Book now
        </span>
        <span className={css(styles.actionAmount)}>
            {price}
        </span>
    </button>
