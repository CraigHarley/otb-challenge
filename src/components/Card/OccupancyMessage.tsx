import React from 'react';

export interface IOccupancyMessageProps {
    adults: number;
    children: number;
    infants: number;
}

export default ({adults, children, infants}: IOccupancyMessageProps) => {
    if (adults && !children && !infants) {
        return (
            <span>
                {adultMessage(adults)}
            </span>
        );
    }

    if (adults && children && infants) {
        return (
            <span>
                {adultMessage(adults)}, {childMessage(children)} & {infantMessage(infants)}
            </span>
        );
    }

    if (adults && children && !infants) {
        return (
            <span>
                {adultMessage(adults)} & {childMessage(infants)}
            </span>
        );
    }

    if (adults && !children && infants) {
        return (
            <span>
                {adultMessage(adults)} & {infantMessage(infants)}
            </span>
        );
    }

    // todo make this impossible with types.
    throw Error("Can't have a holiday with no adults.");
}

const adultMessage = (count: number) => (
    <>
        <strong>{count}</strong> {pluralize('adult', count)}
    </>
);
const childMessage = (count: number) => (
    <>
        <strong>{count}</strong> {pluralizeChildren(count)}
    </>
);
const infantMessage = (count: number) => (
    <>
        <strong>{count}</strong> {pluralize('infant', count)}
    </>
);

// Could also have used https://www.npmjs.com/package/pluralize
// But it's two lines vs a whole dep.
const pluralize = (original: string, count: number): string => count === 1 ? original : original + 's';
const pluralizeChildren = (count: number): string => count === 1 ? 'child' : 'children';
