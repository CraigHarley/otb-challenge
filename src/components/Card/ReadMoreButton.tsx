import React, {ReactElement} from 'react';
import {css, StyleSheet} from "aphrodite";

import config from "../../config";

const styles = StyleSheet.create({
    readMore: {
        position: 'absolute',
        bottom: '0',
        left: '0',
        backgroundColor: 'whitesmoke',
        color: config.styles.colors.primary,
        padding: '0.5rem 1rem',
        fontSize: 'small',
        cursor: 'pointer',
        border: 'none',
        ':hover': {
            color: config.styles.colors.primary,
            backgroundColor: config.styles.colors.secondary
        }
    }
});

export interface IReadMoreButtonProps {
    setIsShowingMoreInfo: Function;
    isShowingMoreInfo: boolean;
}

const ReadMoreButton = ({setIsShowingMoreInfo, isShowingMoreInfo}: IReadMoreButtonProps): ReactElement => (
    <button
        id='btn-read-more'
        className={css(styles.readMore)}
        onClick={() => setIsShowingMoreInfo(!isShowingMoreInfo)}
    >
        <strong>Read more</strong> about this hotel &nbsp;
        {
            isShowingMoreInfo ?
                <i className="fas fa-angle-down"/> :
                <i className="fas fa-angle-right"/>
        }
    </button>
)

export default ReadMoreButton;
