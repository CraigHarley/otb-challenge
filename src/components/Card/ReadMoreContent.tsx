import React, {ReactElement} from 'react';
import {css, StyleSheet} from "aphrodite";

import config from "../../config";

const styles = StyleSheet.create({
    readMoreContent: {
        backgroundColor: 'whitesmoke',
        width: '780px',
        margin: '0 2rem 2rem 2rem',
        padding: '1rem',
        transition: 'transform 0.3s ease-out',
        transform: 'scaleY(1)',
        transformOrigin: 'top',
    },
    h3: {
        fontSize: 'small',
        color: config.styles.colors.primary
    },
    p: {
        fontSize: 'small',
        fontWeight: 'lighter',
        lineHeight: '1.2rem'
    },
    isHidden: {
        transform: 'scaleY(0)',
        height: 0,
        overflow: 'hidden'
    }
});

export interface IReadMoreContentProps {
    isShowingMoreInfo: boolean;
}

export default ({isShowingMoreInfo}: IReadMoreContentProps): ReactElement => (
    <div className={css(styles.readMoreContent, !isShowingMoreInfo && styles.isHidden)}>
        <h3 className={css(styles.h3)}>Overview</h3>
        <p className={css(styles.p)}>
            Lorem Ipsum is simply dummy text of the printing and typesetting industry. Lorem Ipsum has been the
            industry's standard dummy text ever since the 1500s, when an unknown printer took a galley of type
            and
            scrambled it to make a type specimen book. It has survived not only five centuries, but also the
            leap
            into electronic typesetting, remaining essentially unchanged. It was popularised in the 1960s with
            the
            release of Letraset sheets containing Lorem Ipsum passages, and more recently with desktop
            publishing
            software like Aldus PageMaker including versions of Lorem Ipsum.
        </p>
    </div>
)
