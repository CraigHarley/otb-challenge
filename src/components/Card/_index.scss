@import "src/variables";

.content {
  grid-area: Content;

  .card {
    position: relative;
    margin: 2rem 2rem 0 2rem;
    width: 760px;
    height: 300px;
    background-color: whitesmoke;
    display: flex;

    div.content-container {
      padding: 1rem 2rem 2rem 2rem;
      color: $primary;

      h1 {
        font-weight: normal;
        font-size: 1rem;
      }

      h2 {
        font-weight: lighter;
        font-size: 0.8rem;
      }

      .fas {
        color: $secondary;
      }

      ul {
        list-style: none;
        padding-left: 0;
        font-weight: lighter;
        font-size: 0.8rem;
        color: black;
      }

      .call-to-action {
        margin-top: 2rem;
        color: $primary;
        background-color: $secondary;
        width: 100%;
        height: 4rem;
        border: none;
        border-radius: 0.2rem;
        cursor: pointer;

        .action-text {
          font-size: small;
          display: block;
        }

        .action-amount {
          font-weight: bold;
          font-size: medium;
        }

        &:hover {
          color: lighten($primary, 10%);
          background-color: lighten($secondary, 10%);
        }
      }
    }


    // todo - details + summary would be more semantic
    button.read-more {
      position: absolute;
      bottom: 0;
      left: 0;
      background-color: whitesmoke;
      color: $primary;
      padding: 0.5rem 1rem;
      font-size: small;
      cursor: pointer;
      border: none;

      &:hover {
        color: lighten($primary, 10%);
        background-color: lighten(whitesmoke, 10%);
      }
    }

  }

  .read-more-content {
    background-color: whitesmoke;
    width: 760px;
    margin: 0 2rem 2rem 2rem;
    padding: 1rem;
    transition: transform 0.3s ease-out;
    transform: scaleY(1);
    transform-origin: top;

    h3 {
      font-size: small;
      color: $primary;
    }

    p {
      font-size: small;
      font-weight: lighter;
      line-height: 1.2rem;
    }

    &.is-hidden {
      transform: scaleY(0);
      height: 0;
      overflow: hidden;
    }
  }
}
