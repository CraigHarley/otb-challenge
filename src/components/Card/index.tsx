import React, {ReactElement, useState} from 'react';
import {format} from 'date-fns';

import config from '../../config'
import {IHotel, IListable} from '../../types';
import OccupancyMessage from './OccupancyMessage';

export interface ICardProps extends IListable {
    hotel: IHotel;
}

export default ({
                    hotel: {
                        image,
                        name,
                        region,
                        adults,
                        children,
                        infants,
                        price,
                        stars,
                        date
                    }
                }: ICardProps): ReactElement => {

    const [isShowingMoreInfo, setIsShowingMoreInfo] = useState(false);

    return (
        <section>
            <div className="card">
                <img
                    src={`${config.ASSET_URL}${image}`}
                    alt="Image of the hotel"
                />
                <div className="content-container">
                    <h1>{name}</h1>
                    <h2>{region}</h2>
                    <div>
                        {
                            new Array(stars)
                                .fill('')
                                .map((_, key) => <i key={key} className="fas fa-star"/>)
                        }
                    </div>

                    <ul>
                        <li>
                            <OccupancyMessage
                                adults={adults}
                                children={children}
                                infants={infants}
                            />
                        </li>
                        <li><strong>{format(date, 'do MMMM yyyy')}</strong> for <strong>7 days</strong></li>
                        <li>departing from <strong>East Midlands</strong></li>
                    </ul>

                    <button className="call-to-action">
                    <span className="action-text">
                        Book now
                    </span>
                        <span className="action-amount">
                        {price}
                    </span>
                    </button>
                </div>
                <button
                    className="read-more"
                    onClick={() => setIsShowingMoreInfo(!isShowingMoreInfo)}
                >
                    <strong>Read more</strong> about this hotel &nbsp;
                    {
                        isShowingMoreInfo ?
                            <i className="fas fa-angle-down"/> :
                            <i className="fas fa-angle-right"/>
                    }

                </button>
            </div>
            <div className={`read-more-content ${isShowingMoreInfo ? '' : 'is-hidden'}`}>
                <h3>Overview</h3>
                <p>
                    Lorem Ipsum is simply dummy text of the printing and typesetting industry. Lorem Ipsum has been the
                    industry's standard dummy text ever since the 1500s, when an unknown printer took a galley of type
                    and
                    scrambled it to make a type specimen book. It has survived not only five centuries, but also the
                    leap
                    into electronic typesetting, remaining essentially unchanged. It was popularised in the 1960s with
                    the
                    release of Letraset sheets containing Lorem Ipsum passages, and more recently with desktop
                    publishing
                    software like Aldus PageMaker including versions of Lorem Ipsum.
                </p>
            </div>
        </section>

    );
};
