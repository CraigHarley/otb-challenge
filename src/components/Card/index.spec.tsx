import React from 'react';
import {mount, render} from "enzyme";

import Card, {ICardProps} from ".";
import {hotels} from "../../models";

const defaultProps: ICardProps = {
    hotel: hotels[0]
}

test('it should render with sorts',
    () =>
        expect(
            render(
                <Card
                    {...defaultProps}
                />
            )
        )
            .toMatchSnapshot('default')
);

test('when read more is clicked, it should show more content',
    () => {
        const element = mount(
            <Card
                {...defaultProps}
            />
        );

        element
            .find('.read-more')
            .first()
            .simulate('click');

        expect(
            element.render()
        ).toMatchSnapshot('with more content');
    }
);
