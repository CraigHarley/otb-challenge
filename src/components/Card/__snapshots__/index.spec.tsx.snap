// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Card it should render with sorts: default 1`] = `
<section>
  <div
    class="card"
  >
    <img
      alt="Image of the hotel"
      src="http://localhost:9999/hotel-image-1.png"
    />
    <div
      class="content-container"
    >
      <h1>
        Iberostar Grand Salome
      </h1>
      <h2>
        Costa Adeje, Tenerife
      </h2>
      <div>
        <i
          class="fas fa-star"
        />
        <i
          class="fas fa-star"
        />
        <i
          class="fas fa-star"
        />
        <i
          class="fas fa-star"
        />
      </div>
      <ul>
        <li>
          <span>
            <strong>
              2
            </strong>
             adults & 
            <strong>
              1
            </strong>
             infant
          </span>
        </li>
        <li>
          <strong>
            3rd July 2019
          </strong>
           for 
          <strong>
            7 days
          </strong>
        </li>
        <li>
          departing from 
          <strong>
            East Midlands
          </strong>
        </li>
      </ul>
      <button
        class="call-to-action"
      >
        <span
          class="action-text"
        >
          Book now
        </span>
        <span
          class="action-amount"
        >
          £3,136.50
        </span>
      </button>
    </div>
    <button
      class="read-more"
    >
      <strong>
        Read more
      </strong>
       about this hotel  
      <i
        class="fas fa-angle-right"
      />
    </button>
  </div>
  <div
    class="read-more-content is-hidden"
  >
    <h3>
      Overview
    </h3>
    <p>
      Lorem Ipsum is simply dummy text of the printing and typesetting industry. Lorem Ipsum has been the industry's standard dummy text ever since the 1500s, when an unknown printer took a galley of type and scrambled it to make a type specimen book. It has survived not only five centuries, but also the leap into electronic typesetting, remaining essentially unchanged. It was popularised in the 1960s with the release of Letraset sheets containing Lorem Ipsum passages, and more recently with desktop publishing software like Aldus PageMaker including versions of Lorem Ipsum.
    </p>
  </div>
</section>
`;

exports[`Card when read more is clicked, it should show more content: with more content 1`] = `
<section>
  <div
    class="card"
  >
    <img
      alt="Image of the hotel"
      src="http://localhost:9999/hotel-image-1.png"
    />
    <div
      class="content-container"
    >
      <h1>
        Iberostar Grand Salome
      </h1>
      <h2>
        Costa Adeje, Tenerife
      </h2>
      <div>
        <i
          class="fas fa-star"
        />
        <i
          class="fas fa-star"
        />
        <i
          class="fas fa-star"
        />
        <i
          class="fas fa-star"
        />
      </div>
      <ul>
        <li>
          <span>
            <strong>
              2
            </strong>
             adults & 
            <strong>
              1
            </strong>
             infant
          </span>
        </li>
        <li>
          <strong>
            3rd July 2019
          </strong>
           for 
          <strong>
            7 days
          </strong>
        </li>
        <li>
          departing from 
          <strong>
            East Midlands
          </strong>
        </li>
      </ul>
      <button
        class="call-to-action"
      >
        <span
          class="action-text"
        >
          Book now
        </span>
        <span
          class="action-amount"
        >
          £3,136.50
        </span>
      </button>
    </div>
    <button
      class="read-more"
    >
      <strong>
        Read more
      </strong>
       about this hotel  
      <i
        class="fas fa-angle-down"
      />
    </button>
  </div>
  <div
    class="read-more-content "
  >
    <h3>
      Overview
    </h3>
    <p>
      Lorem Ipsum is simply dummy text of the printing and typesetting industry. Lorem Ipsum has been the industry's standard dummy text ever since the 1500s, when an unknown printer took a galley of type and scrambled it to make a type specimen book. It has survived not only five centuries, but also the leap into electronic typesetting, remaining essentially unchanged. It was popularised in the 1960s with the release of Letraset sheets containing Lorem Ipsum passages, and more recently with desktop publishing software like Aldus PageMaker including versions of Lorem Ipsum.
    </p>
  </div>
</section>
`;
